### ISSUE TITLE
JUnit reporter testsuite counts are not consistent with number of testcases

### ISSUE THREAD

**Issue by michaelbprice** [philsquared/Catch#529]
The test code:

```c++
TEST_CASE("Single TEST_CASE", "single")
{
    CHECK(true);
    CHECK(true);
    CHECK(false);
}
```

produces the following JUnit reporter output:

```xml
<testsuites>
  <testsuite name="all tests" errors="0" failures="1" tests="3" hostname="tbd" time="1.9e-05" timestamp="tbd">
    <testcase classname="global" name="Single TEST_CASE" time="0.000011">
      <failure message="false" type="CHECK">
at /file_path_redacted/bug.cpp:14
      </failure>
    </testcase>
    <system-out/>
    <system-err/>
  </testsuite>
</testsuites>
```

I'd expect the number of `testcase` elements to match the value of the `tests` attribute on the `testsuite` element.  It appears that the attribute is instead counting the number of individual assertions that are made.  I believe that is misleading behavior (and unfortunately, I'm not sure the semantics of the JUnit format are terribly clear).

**Comment by michaelbprice** [comment]
Another, quite worse example (from the code that we *actually* had in some of our tests):

```c++
TEST_CASE("Must throw an integer", "throw")
{
    try
    {
        throw 1;
        FAIL("Did not throw");
    }
    catch (int)
    {
        // This is expected to happen
    }
}
```

Output:

```xml
<testsuites>
  <testsuite name="all tests" errors="0" failures="0" tests="0" hostname="tbd" time="4.4e-05" timestamp="tbd">
    <system-out/>
    <system-err/>
  </testsuite>
</testsuites>
```

We've since replaced this with an EXPECTS assertion.  I think that writing tests like above seems a valid use of the framework, but it think it creates a quite confusing situation given the output.

**Comment by martinmoene** [comment]
[Assertion Macros, Exceptions](https://github.com/philsquared/Catch/blob/master:/docs/assertions.md#exceptions):
```
REQUIRE_THROWS( throw 1 );
REQUIRE_THROWS_AS( throw std::runtime_error("blow"), std::runtime_error );
```

**Comment by michaelbprice** [comment]
@martinmoene Yes, those are the ones that I meant when I mentioned `EXPECTS` assertions.  Just got the macro name wrong.  But, I do believe that the second example should report correctly.  It possible that when taking different branches of control flow, that a single test case might have differing number of assertions.

**Comment by philsquared** [comment]
@michaelbprice I'm not sure what the problem is with your second example (with the `throw 1`) - is it that no assertions are reported - even though there is a `FAIL` written there? It is not possible for Catch to even be aware of that `FAIL`, given that it never reaches that code - and even if it could (perhaps some form of reflection) I don't think there is any logically consistent way it could be reported - or am I missing something?

As for the original issue I think the key here is, "I'm not sure the semantics of the JUnit format are terribly clear". The JUnit/Ant format is not documented so is open to interpretation. I think what Catch is doing here is a valid interpretation. IIRC I made this choice because there was no other way to get the individual number of assertions reported.

**Comment by michaelbprice** [comment]
>@michaelbprice I'm not sure what the problem is with your second example (with the throw 1) - is it that no assertions are reported - even though there is a FAIL written there? It is not possible for Catch to even be aware of that FAIL, given that it never reaches that code - and even if it could (perhaps some form of reflection) I don't think there is any logically consistent way it could be reported - or am I missing something?

It is that the `tests` attribute is `0`, even though I clearly have a test defined.  Basically, I feel you are incorrectly/misleadingly keying the value of that attribute off of the number of assertion macros that are executed, instead of the number of `TEST_CASE` macros that exist.

I'll do an experiment later to see how some Java tools behave with regards to counting the total number of tests.

>As for the original issue I think the key here is, "I'm not sure the semantics of the JUnit format are terribly clear". The JUnit/Ant format is not documented so is open to interpretation. I think what Catch is doing here is a valid interpretation. IIRC I made this choice because there was no other way to get the individual number of assertions reported.

Hopefully my response above clarified.  I don't think you should be trying to count individual assertions, but rather the number of test cases.

**Comment by michaelbprice** [comment]
As another example, if I were designing an interface such that misuses are diagnosed at compile-time rather than run-time, I would very much like to do something like this:

```c++
#include <type_traits>

TEST_CASE("Class X must be default constructible", "static_tests")
{
    static_assert(std::is_default_constructible_v<X>); // tentatively valid C++17
}
```

In that case, I'd expect any failures to be reported at compile time, but I'd also like the unit test framework to affirmatively report that the test case is good (at least by recognizing that it is in fact a test!)

I could hack around it by adding something goofy like a Catch assertion that just asserts on true, but I seriously shouldn't have to do something like that for every place where I want to do this (possibly dozens on any one type in my library).

**Comment by michaelbprice** [comment]
As an additional data point... Jenkins agrees with my perspective...

This XML file:

```xml
<?xml version="1.0" encoding="utf-8"?>
<testsuites>
  <testsuite name="all tests" errors="0" failures="0" tests="26" hostname="tbd" time="0.004012" timestamp="tbd">
    <testcase classname="global" name="Load this library" time="0.000265"/>
    <testcase classname="global" name="Load Dummy library" time="0.000468"/>
    <testcase classname="global" name="Load Dummy library twice" time="0.000433"/>
    <testcase classname="global" name="Load Nonexistant library" time="0.000266"/>
    <testcase classname="global" name="Load Nonexistant function from Dummy library" time="0.000408"/>
    <testcase classname="global" name="DynamicLibraryLoader is not default constructable" time="0.000057"/>
    <testcase classname="global" name="Get last path separator position" time="0.000051"/>
    <testcase classname="global" name="Get last path separator position with no path separator" time="0.000033"/>
    <testcase classname="global" name="Get extension position with path separator" time="0.000045"/>
    <testcase classname="global" name="Get extension position without path separator" time="0.000031"/>
    <testcase classname="global" name="Strip directories with path separator" time="0.00003"/>
    <testcase classname="global" name="Strip directories without path separator" time="0.000066"/>
    <testcase classname="global" name="Strip extension with path separator" time="0.000033"/>
    <testcase classname="global" name="Strip extension without path separator" time="0.000029"/>
    <testcase classname="global" name="Strip directories and extension with path separator" time="0.000057"/>
    <testcase classname="global" name="Strip directories and extension without path separator" time="0.000057"/>
    <testcase classname="global" name="Get destination filename" time="0.000188"/>
    <testcase classname="global" name="Adds XML declaration if it is missing" time="0.000242"/>
    <testcase classname="global" name="No duplicate XML declaration if one already existed" time="0.000086"/>
    <system-out/>
    <system-err/>
  </testsuite>
</testsuites>
```

is reported as:

![image](https://cloud.githubusercontent.com/assets/60362/11157427/f03196d8-8a16-11e5-84ca-5079f26d0ea7.png)
